wrapper/src/main/java/software/amazon/jdbc/util/storage/StorageServiceImpl.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.util.storage;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import software.amazon.jdbc.plugin.customendpoint.CustomEndpointInfo;
+import software.amazon.jdbc.util.ItemDisposalFunc;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.ShouldDisposeFunc;
+
+public class StorageServiceImpl implements StorageService {
+  private static final Logger LOGGER = Logger.getLogger(StorageServiceImpl.class.getName());
+  protected static final long DEFAULT_CLEANUP_INTERVAL_NANOS = TimeUnit.MINUTES.toNanos(1);
+  protected static final Map<String, ExpirationCache<Object, ?>> caches = new ConcurrentHashMap<>();
+  protected static final Map<String, Supplier<ExpirationCache<Object, ?>>> defaultCacheSuppliers;
+  protected static final AtomicBoolean isInitialized = new AtomicBoolean(false);
+  protected static final ReentrantLock initLock = new ReentrantLock();
+  protected static final Map<String, Long> cleanupTimes = new ConcurrentHashMap<>();
+  protected static final ScheduledExecutorService cleanupExecutor = Executors.newSingleThreadScheduledExecutor((r -> {
+    final Thread thread = new Thread(r);
+    thread.setDaemon(true);
+    return thread;
+  }));
+
+  static {
+    Map<String, Supplier<ExpirationCache<Object, ?>>> suppliers = new HashMap<>();
+    suppliers.put(ItemCategory.TOPOLOGY, () -> new ExpirationCacheBuilder<>(Topology.class).build());
+    suppliers.put(ItemCategory.CUSTOM_ENDPOINT, () -> new ExpirationCacheBuilder<>(CustomEndpointInfo.class).build());
```

Comment:
This type of data might be not used by user, in a particular driver configuration. I think it should be initialized by corresponding plugin

Reply:
This line does not actually instantiate any caches, it just defines a lambda that can be used to create the custom endpoint info cache if it is needed. The lambda is only ever called the first time someone tries to store something in the custom endpoint cache, at which point it will be created. The benefit of this is that our code doesn't need to worry about whether the cache exists or not yet. Instead, it will be automatically created, but only if it is actually needed.